domains/certificates/Query.API/API/Query.API/Controllers/ContractsController.cs

```diff
@@ -45,21 +48,70 @@ public class ContractsController : ControllerBase
             return ValidationProblem(ModelState);
         }
 
-        var startDate = DateTimeOffset.FromUnixTimeSeconds(createContract.StartDate);
-        DateTimeOffset? endDate = createContract.EndDate.HasValue ? DateTimeOffset.FromUnixTimeSeconds(createContract.EndDate.Value) : null;
+        var startDate = UnixTimestamp.Create(createContract.StartDate);
+        UnixTimestamp? endDate = createContract.EndDate.HasValue
+            ? UnixTimestamp.Create(createContract.EndDate.Value)
+            : null;
 
         var result = await service.Create(
-            createContract.GSRN,
+            [(createContract.GSRN, startDate, endDate)],
             user,
-            startDate,
-            endDate,
             cancellationToken);
 
         return result switch
         {
-            GsrnNotFound => ValidationProblem($"GSRN {createContract.GSRN} not found"),
+            GsrnNotFound => ValidationProblem(),
             ContractAlreadyExists => ValidationProblem(statusCode: 409),
-            CreateContractResult.Success(var createdContract) => CreatedAtRoute("GetContract", new { id = createdContract.Id }, Contract.CreateFrom(createdContract)),
+            CreateContractResult.Success(var createdContract) => CreatedAtRoute("GetContract",
+                new { id = createdContract[0].Id }, Contract.CreateFrom(createdContract[0])),
+            _ => throw new NotImplementedException($"{result.GetType()} not handled by {nameof(ContractsController)}")
+        };
+    }
+
+    /// <summary>
+    /// Create contracts that activates granular certificate generation for a metering point
+    /// </summary>
+    [HttpPost]
+    [ProducesResponseType(201)]
+    [ProducesResponseType(400)]
+    [ProducesResponseType(typeof(void), 409)]
+    [ApiVersion(ApiVersions.Version20240423)]
+    [Route("api/certificates/contracts")]
+    public async Task<ActionResult> CreateContract(
+        [FromBody] CreateContracts createContracts,
+        [FromServices] IValidator<CreateContract> validator,
+        [FromServices] IContractService service,
+        CancellationToken cancellationToken)
+    {
+        var user = new UserDescriptor(User);
+        var certificateIssuingContracts = new List<(string gsrn, UnixTimestamp startDate, UnixTimestamp? endDate)>();
```

Comment:
Couldn't "(string gsrn, UnixTimestamp startDate, UnixTimestamp? endDate)" be a class / record?

Reply:
fixed